### PR TITLE
BFD-3064: Updated test and PatientProviderResource classes IRT 404 err.

### PR DESCRIPTION
**Pull request by aadedejifearless**
**JIRA Ticket:**
[BFD-3064](https://jira.cms.gov/browse/BFD-3064)

**User Story or Bug Summary:**

BCDA is generating a lot of 404's with the following error messages

failure during REST processing: ca.uhn.fhir.rest.server.exceptions.ResourceNotFoundException: By hash query found more than one distinct BENE_ID: 2. This change adds a distinctBeneIdsList so we can debug the actual  Bene Ids that are creating the error. 

---

### What Does This PR Do?

- Updates the `apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java`.
- Updates `apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java`.
- Updates both the `testPatientByIdentifierWhenMbiHashWithBeneDupesExpect404()` and `testPatientByIdentifierWhenMbiHashWithBeneDupesExpect404()` in `apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/PatientE2EBase.java`.
-  Updates both the `testPatientByIdentifierWhenMbiHashWithBeneDupesExpect404()` and `testPatientByIdentifierWhenMbiHashWithBeneDupesExpect404()` in `apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/stu3/providers/PatientE2E.java`

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [X] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [X] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [X] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [X] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [X] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
*  [X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X] The data dictionary has been updated with any field mapping changes, if any were made.
* [X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [X] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java [hunk 1]
```diff
@@ -906,22 +907,31 @@ private Patient queryDatabaseByHash(
...
     }
-
     // Then, if we found more than one distinct BENE_ID, or none, throw an error.
     long distinctBeneIds =
         matchingBenes.stream()
             .map(Beneficiary::getBeneficiaryId)
             .filter(Objects::nonNull)
             .distinct()
             .count();
+    List<Long> distinctBeneIdsList =
```
Comment (mackec1) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java:917]: couldn't we move this down to where we've determined that the distinctBeneIds > 1 ?

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java [hunk 2]
```diff
@@ -938,14 +939,24 @@ private Patient queryDatabaseByHash(
             .filter(Objects::nonNull)
             .distinct()
             .count();
+    List<Long> distinctBeneIdsList =
```
Comment (mackec1) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java:942]: ditto for comment from the R4 provier.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java [hunk 3]
```diff
@@ -906,22 +907,31 @@ private Patient queryDatabaseByHash(
...
           matchingBenes.size());
     }
-
     // Then, if we found more than one distinct BENE_ID, or none, throw an error.
     long distinctBeneIds =
         matchingBenes.stream()
             .map(Beneficiary::getBeneficiaryId)
             .filter(Objects::nonNull)
             .distinct()
             .count();
```
Comment (billie-rose) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java:916]: we should simplify this, we don't need to have two lists here

```java
List<Long> distinctBeneIds =
        matchingBenes.stream()
            .map(Beneficiary::getBeneficiaryId)
            .distinct()
            .collect(Collectors.toList());

    if (distinctBeneIds.size() == 0) {
      throw new NoResultException();
    } else if (distinctBeneIdsList.size() > 1) {
       BfdMDC.put(
          "database_query_by_hash_collision_distinct_bene_ids", Long.toString(distinctBeneIds));
      throw new ResourceNotFoundException(
          "By hash query found more than one distinct BENE_ID: "
// Guessing you meant to put the size here and not the list twice
              + distinctBeneIds.size()
              + ", DistinctBeneIdsList: "
              + distinctBeneIdsList);
   }

      Beneficiary beneficiary = matchingBenes.get(0);
```

Not sure if `.filter(Objects::nonNull)` is relevant when it's placed after `.map(Beneficiary::getBeneficiaryId)`. Would drop the sorting as well, personally, for performance reasons

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java [hunk 4]
```diff
@@ -906,22 +907,31 @@ private Patient queryDatabaseByHash(
...
     Beneficiary beneficiary = null;
     if (distinctBeneIds <= 0) {
       throw new NoResultException();
     } else if (distinctBeneIds > 1) {
       BfdMDC.put(
           "database_query_by_hash_collision_distinct_bene_ids", Long.toString(distinctBeneIds));
       throw new ResourceNotFoundException(
-          "By hash query found more than one distinct BENE_ID: " + Long.toString(distinctBeneIds));
+          "By hash query found more than one distinct BENE_ID: "
+              + Long.toString(distinctBeneIds)
```
Comment (billie-rose) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java:932]: should this be the size?

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java [hunk 5]
```diff
@@ -19,6 +19,7 @@
 import com.codahale.metrics.MetricRegistry;
 import com.codahale.metrics.Timer;
 import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableList;
```
Comment (billie-rose) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java:22]: Do we need to use this over `Collectors.toList()`?

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/PatientE2EBase.java [hunk 6]
```diff
@@ -160,9 +162,23 @@ public void testPatientByIdentifierWhenMbiHashWithBeneDupesExpect404() {
...
-    // basically second param doesnt matter for this test
+    // basically second param doesn't matter for this test
     String mbiHash = getMbiHash(currentMbi, false, loadedRecords);
 
+    Map<Optional<String>, List<Beneficiary>> beneficiaryMap =
+        loadedRecords.stream()
+            .filter(Beneficiary.class::isInstance)
+            .map(Beneficiary.class::cast)
+            .toList()
+            .stream()
```
Comment (billie-rose) [inline apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/PatientE2EBase.java:173]: Why `toList()` and then immediately `stream()`?

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/PatientE2EBase.java [hunk 7]
```diff
@@ -160,9 +162,23 @@ public void testPatientByIdentifierWhenMbiHashWithBeneDupesExpect404() {
...
+            .collect(Collectors.groupingBy(Beneficiary::getMbiHash));
+
+    List<Long> distinctBeneIdList =
```
Comment (billie-rose) [inline apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/PatientE2EBase.java:176]: Wouldn't this need `.distinct()` like its implementor ?

**Comment by billie-rose** [comment]
Some comments, most of them apply to places where the commented code is duplicated even if I didn't specify it. Thank you!

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java (hunk 4 above)
Comment (billie-rose) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java:932]: nvm

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java (hunk 2 above)
Comment (aadedejifearless) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java:942]: Same as the previous comment.

**Review by mackec1 — DISMISSED** [review]
Looks good!

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java (hunk 5 above)
Comment (aadedejifearless) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java:22]: Immutable Collections provide:
Less heap space: The space required to store a collection data is very less as compared with the traditional approach in earlier versions of java.
Faster access to data: As the overhead to store data and wrap into Collections.unmodifiable is reduced, now data access becomes faster.

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/PatientE2EBase.java [hunk 8]
```diff
@@ -177,7 +193,12 @@ public void testPatientByIdentifierWhenMbiHashWithBeneDupesExpect404() {
         .expect()
         .statusCode(404)
         .body("issue.severity", hasItem("error"))
-        .body("issue.diagnostics", hasItem("By hash query found more than one distinct BENE_ID: 5"))
+        .body(
+            "issue.diagnostics",
+            hasItem(
+                "By hash query found more than one distinct BENE_ID: 5,"
```
[the remark this answers is not in the document]
Comment (aadedejifearless) [reply apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/PatientE2EBase.java:199]: Changed.

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/PatientE2EBase.java (hunk 6 above)
Comment (aadedejifearless) [reply apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/PatientE2EBase.java:173]: Removd this. Not needed

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java (hunk 3 above)
Comment (aadedejifearless) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java]: Removed the .filter(Objects::nonNull)

**Review by mackec1 — APPROVED** [review]

**Pull request merged**